spec/integration/header_parser_spec.rb

```diff
@@ -15,26 +16,102 @@ def to_dto
           end
         end
       end
+
       context 'without headers' do
         let(:headers) { nil }
         it { expect { to_dto }.to raise_error(RequestHandler::MissingArgumentError) }
       end
+
       context 'with headers set correctly' do
         let(:headers) { { 'HTTP_APP_KEY' => 'some.app.key', 'HTTP_USER_ID' => '345' } }
         it { expect(to_dto).to eq(OpenStruct.new(headers: { app_key: 'some.app.key', user_id: '345' })) }
       end
     end
   end
 
+  shared_examples 'it validates the headers' do
+    subject(:to_dto) { class_with_headers_schema.new(request: request).to_dto }
+    let(:request) { build_mock_request(params: {}, headers: headers, body: '') }
+
+    context 'HeaderParser' do
+      let(:class_with_headers_schema) do
+        Class.new(RequestHandler::Base) do
+          options do
+            headers do
+              schema(Dry::Schema.Params do
+                required(:client_id).filled(:string)
+              end)
+            end
+          end
+
+          def to_dto
+            OpenStruct.new(
+              headers: headers
+            )
+          end
+        end
+      end
+
+      context 'when the headers are invalid' do
```

Comment:
headers always come in as string (unless rack or the web framework are doing something wrong), so I'd change this to a more realistic example. E.g doing exactly the opposite by having a coercible int and then expecting it to fail on a non-decimal string.